File: tiq-e/trees/4/main.ts
import { performance } from "perf_hooks";

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function levelOrder(root: TreeNode | null): number[][] {
  return bfs([root], [], 0);
}

function bfs(
  que: Array<TreeNode | null>,
  res: number[][],
  depth: number
): number[][] {
  if (!que.length) return res;
  const newQ: Array<TreeNode | null> = [];
  que.forEach((node: TreeNode | null) => {
    if (node) {
      if (!res[depth]) res[depth] = [];
      res[depth].push(node.val);
      newQ.push(node.left);
      newQ.push(node.right);
    }
  });
  return bfs(newQ, res, depth + 1);
}

const root = new TreeNode(5, null, null);

const start = performance.now();
const result = levelOrder(root);
const end = performance.now();
console.log("Duration:", end - start);
console.log("Result:", result);

File: tiq-e/trees/4/main.rs
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
use std::time::Instant;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut queue: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        let mut ans: Vec<Vec<i32>> = Vec::new();

        if let Some(node) = root {
            queue.push_back(node);
        } else {
            return ans;
        }

        while !queue.is_empty() {
            ans.push(queue.iter().map(|node| node.borrow().val).collect());
            for _ in 0..queue.len() {
                if let Some(node) = queue.pop_front() {
                    if let Some(left) = node.borrow().left.clone() {
                        queue.push_back(Rc::clone(&left));
                    }
                    if let Some(right) = node.borrow().right.clone() {
                        queue.push_back(Rc::clone(&right));
                    }
                }
            }
        }
        ans
    }
}

fn main() {
    let root = TreeNode(5, None, None);

    let start = Instant::now();
    let result = Solution::level_order(root);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
